Реализация микросервиса "Отчетность"

r4p1er opened the pull request:
TL;DR: реализовал микросервис "Отчетность" (Reporting), всего 3 сборки: Reporting.Domain со всей бизнес логикой, Reporting.Infrastructure с подключением к БД и реализацией репозитория отчетов, Reporting.Api с RESTful API микросервиса

---

Все сборки реализованы так же, как и в других микросервисах. В том числе присутствует валидация входящих данных с помощью библиотеки FluentValidation

Список API микросервиса Reporting:
 * GET api/reports - Получение своих отчетов персоналом
 * GET api/reports/{id} - Получение отчета по идентификатору
 * POST api/reports - Создание отчета за период
 * DELETE api/reports/{id} - Удаление отчета

~~UPD: в сборку Hotel.Shared добавил кастомные сервисы по работе с брокером сообщений RabbitMQ. Сервис соединения с RabbitMQ добавляется как singleton (т.к. в рамках одного микросервиса достаточно только одного соединения), а основной сервис, `IRabbitService`, добавляется как transient, т.к. я специально спроектировал его таким образом, чтобы его использование предполагало собой создание каждый раз нового экземпляра. Если использовать его иначе, то это может привести к потенциальным ошибкам.~~

~~Сервис `IReportService` обновил таким образом, чтобы отчет создавался автоматически, путем запрашивания данных через сервис `IRabbitService` у других микросервисов. В сборке Reporting.Api зарегистрировал все сервисы, связанные с RabbitMQ.~~

~~В микросервисах Booking и Managing зарегистрировал сервисы, связанные с RabbitMQ, а также добавил hosted сервисы `RabbitHostedService`, для фонового прослушивания входящих сообщений.~~

UPD2: переписал все под библиотеку MassTransit вместо RabbitMQ.Client.

r4p1er committed:
Added Reporting.Domain assembly; created as other domain assemblies; validators included

r4p1er committed:
Added validation for the JSON field Data in the ReportData object

r4p1er committed:
Added Reporting.Infrastructure assembly; created as other infrastructure assemblies

r4p1er committed:
Created ReportDTO representing data field as JsonDocument; ReportData was changes so data field is also JsonDocument

r4p1er committed:
Added Reporting.Api assembly created as other api assemblies; created InitialMigration

SyReX17 commented on Reporting/Reporting.Api/Controllers/ReportsController.cs:
```diff
@@ -0,0 +1,48 @@
+
+[ApiController]
+[Route("api/[controller]")]
+public class ReportsController : ControllerBase
+{
+    private readonly IReportService _reportService;
+
+    public ReportsController(IReportService reportService)
+    {
+        _reportService = reportService;
+    }
```
Я не помню писал или нет, но в .net 8 tcnm primary конструкторы, и их можно использовать

SyReX17 commented on Reporting/Reporting.Api/Middlewares/ErrorHandlingMiddleware.cs:
```diff
@@ -0,0 +1,41 @@
+
+namespace Reporting.Api.Middlewares;
+
+public class ErrorHandlingMiddleware
+{
+    private readonly RequestDelegate _next;
+
+    public ErrorHandlingMiddleware(RequestDelegate next)
+    {
+        _next = next;
+    }
+    
+    public async Task InvokeAsync(HttpContext context)
+    {
+        try
+        {
+            await _next.Invoke(context);
+        }
+        catch (Exception e)
+        {
+            await HandleExceptionAsync(context, e);
+        }
+    }
+
+    private async Task HandleExceptionAsync(HttpContext context, Exception exception)
+    {
+        context.Response.ContentType = "application/json";
+        context.Response.StatusCode = exception is RequestException
+            ? (int)(exception as RequestException)!.StatusCode
+            : StatusCodes.Status500InternalServerError;
+        var result = new
+        {
+            message = context.Response.StatusCode == StatusCodes.Status500InternalServerError
+                ? "Internal server error"
+                : exception.Message
+        };
+
+        await context.Response.WriteAsJsonAsync(result);
+    }
+}
```
Это же во всех микросервисах используется, можно создать отдельную сборку Hotel.Shared и туда вынести все, что переиспользуется (методы расширения, исключения и т.д.)

SyReX17 commented on Reporting/Reporting.Domain/Services/ReportService.cs:
```diff
@@ -0,0 +1,62 @@
+        return new ReportDTO(report);
+    }
+
+    public async Task<ReportDTO> CreateReport(ReportData data)
+    {
+        await _validator.ValidateAndThrowAsync(data);
+
+        var report = new Report()
+        {
+            Id = Guid.NewGuid(),
+            Summary = data.Summary,
+            From = data.From,
+            To = data.To,
+            Data = JsonSerializer.Serialize(data.Data.RootElement)
+        };
+
+        await _repository.AddReportAsync(report);
+
+        return new ReportDTO(report);
+    }
```
Я видимо немного не так отписал в задании, создание отчета означает, что сам сервис должен сформировать отчет, у тебя приходят (summary, from, to), а данные по занятости нужно запросить из других сервисов

r4p1er committed:
Refactoring: use primary constructor instead of usual

r4p1er committed:
Refactoring: ErrorHandlingMiddleware is moved to the Hotel.Shared assembly. Startup classes use primary constructors

r4p1er committed:
Feat: add RabbitMQ communication logic to the Hotel.Shared assembly

Add an interface and its implementation of RabbitMQ logic, options for the service, message model, extensions to register the service

r4p1er committed:
Refactoring: refactor AddMessageHandler method of RabbitService

Handler must return bool whether it fully processed message or not

r4p1er committed:
Refactoring: refactor logic of communication with the RabbitMQ

Register RabbitMQ connection as singleton. Register IRabbitService as transient. Add ResponseTarget prop to the RabbitMessage model. Completely refactor RabbitService

r4p1er committed:
Feat: create report data via data provided by other services

And other minor fixes

r4p1er committed:
Feat: add shared services and their options

Add services and options to the service collection of the Reporting.Api assembly

r4p1er committed:
Feat: add rabbit services and options to the Managing microservice

r4p1er committed:
Feat: adds configuration and RabbitMQ integration for the Booking API. This commit enables the use of a RabbitMQ queue in the application, allowing for more efficient communication between services. It also adds configuration options to customize the application's behavior

r4p1er committed:
Refactoring: update RabbitHostedService to use IServiceScopeFactory instead of IRoomService and inject service scope

r4p1er committed:
Refactoring: update ReportService: Correctly filter bookings by date range

SyReX17 commented on Hotel.Shared/Services/RabbitService.cs:
```diff
@@ -0,0 +1,134 @@
+
+namespace Hotel.Shared.Services;
+
+public class RabbitService : IRabbitService
+{
+    private readonly IModel _channel;
+    private readonly RabbitOptions _options;
+    private string _consumerTag = string.Empty;
+    private readonly JsonSerializerOptions _serializerOptions = new() { PropertyNameCaseInsensitive = true };
+    
+    public RabbitService(IRabbitConnectionService connection, RabbitOptions options)
+    {
+        _channel = connection.CreateChannel();
+        _options = options;
+
+        _channel.QueueDeclare(
+            queue: options.Queue,
+            durable: false,
+            exclusive: false,
+            autoDelete: false,
+            arguments: null
+        );
+    }
```
RabbitMQ.Client в целом классная либа, но достаточно низкоуровневая, есть более высокоуровневые альтернативы, в частности Masstransit, лучше ее использовать, там уже есть готовые consumer'ы, а publisher'ы сделать не сложно

SyReX17 commented on Managing/Managing.Infrastructure/HostedServices/RabbitHostedService.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace Managing.Infrastructure.HostedServices;
+
+public class RabbitHostedService(IRabbitService rabbit, IServiceScopeFactory scopeFactory) : BackgroundService
```
В целом, если использовать masstransit, то этот класс уйдет, но это еще один пример класса, который можно выносить в shared проекты, так как он несколько раз встречается

SyReX17 commented on Managing/Managing.Api/Startup.cs:
```diff
@@ -1,29 +1,25 @@
-        
+        collection.AddAuth(configuration["Auth:Key"]!);
+
+        collection.AddServicesOptions(configuration);
```
Опции лучше передавать как параметры при регистрации сервисов для которых они применяются, и регистрировать их вместе с самими сервисами, тут несколько моментов при регистрации опций отдельно от сервисов

1. Можно не зарегистрировать опции, но зарегистрировать сервис, тогда будет ошибка
2. Могут быть опциональные сервисы, например, сервисы для кэширования, в таких случаях нужно отличать добавление в DI

Для решения этих проблем можно делать методы расширения под сервис, например AddDatabase(options), AddRabbitMq(options), AddCaching(options), и тому подобное

r4p1er committed:
Refactoring: use MassTransit library for communication with RabbitMq instead of RabbitMQ.Client

r4p1er committed:
Refactoring: register services together with their options

Register Identity services together with their options, not separately
